graf2d/gpad/v7/inc/ROOT/TDrawingOptsBase.hxx

```diff
@@ -0,0 +1,304 @@
+   /// \returns the reference to `val`, which might be `IsInvalid()` if `val` is not part of this table.
+   TOptsAttrRef<PRIMITIVE> SameAs(const PRIMITIVE &val);
+
+   /// Access to the underlying attribute table (non-const version).
+   std::vector<value_type> &GetTable() { return fTable; }
+
+   /// Access to the underlying attribute table (const version).
+   const std::vector<value_type> &GetTable() const { return fTable; }
+};
+
+extern template class TOptsAttrTable<TColor>;
+extern template class TOptsAttrTable<long long>;
+extern template class TOptsAttrTable<double>;
+} // namespace Internal
+
+/** \class ROOT::Experimental::TDrawingOptsBase
+  Base class for drawing options. Implements access to the default and the default's initialization
+  from a config file.
+
+  Drawing options are made up of three kinds of primitives:
+    - TColor
+    - integer (long long)
+    - floating point (double)
+  They register the primitives with their `TCanvas`, and then refer to the registered primitives.
+  Upon destruction, the `TDrawingOptsBase` deregisters the use of its primitives with the `TCanvas`,
+  which empties the respective slots in the tables of the `TCanvas`, unless other options reference
+  the same primitives (through `SameAs()`), and until the last use has deregistered.
+
+  In derived classes (e.g. drawing options for the class `MyFancyBox`), declare attribute members as
+     TOptsAttrRef<TColor> fLineColor{*this, "MyFancyBox.LineColor"};
+  The attribute's value will be taken from the  will be initialized
+  */
+
+class TDrawingOptsBaseNoDefault {
+public:
+   template <class PRIMITIVE>
+   class OptsAttrRefArr {
+      /// Indexes of the `TCanvas`'s attribute table entries used by the options object.
+      std::vector<TOptsAttrRef<PRIMITIVE>> fRefArray;
+
+   public:
+      ~OptsAttrRefArr();
+      /// Register an attribute.
+      ///\returns the index of the new attribute.
+      TOptsAttrRef<PRIMITIVE> Register(TCanvas &canv, const PRIMITIVE &val);
+
+      /// Re-use an existing attribute.
+      ///\returns the index of the attribute (i.e. valRef).
+      TOptsAttrRef<PRIMITIVE> SameAs(TCanvas &canv, TOptsAttrRef<PRIMITIVE> idx);
+
+      /// Re-use an existing attribute.
+      ///\returns the index of the attribute, might be `IsInvalid()` if `val` could not be found.
+      TOptsAttrRef<PRIMITIVE> SameAs(TCanvas &canv, const PRIMITIVE &val);
+
+      /// Update the attribute at index `idx` to the value `val`.
+      void Update(TCanvas &canv, TOptsAttrRef<PRIMITIVE> idx, const PRIMITIVE &val);
+
+      /// Clear all attribute references, removing their uses in `TCanvas`.
+      void Release(TCanvas &canv);
+
+      /// Once copied, elements of a OptsAttrRefArr need to increase their use count.
+      void RegisterCopy(TCanvas &canv);
+   };
+
+private:
+   /// The `TCanvas` holding the `TDrawable` (or its `TPad`).
+   TCanvas *fCanvas; //! always != nullptr
```

Comment:
Would ```TCanvas &fCanvas;``` be an option?

Reply:
Not for the I/O, right?

Reply:
But your marked it transient :)

Reply:
Woops. That was meant to be `//->`!

Reply:
Really spelt that way :) that will be confusing for the reader :) ... (and besides since the class does not have a default constructor the I/O will be challenging)